alacritty_terminal/src/input.rs

```diff
@@ -314,7 +317,7 @@ impl Default for Action {
 
 impl Action {
     #[inline]
-    fn execute<A: ActionContext>(&self, ctx: &mut A, mouse_mode: bool) {
+    fn execute<A: ActionContext>(&self, ctx: &mut A, mouse_mode: bool) -> bool {
```

Comment:
I feel testing for `Action::ReceiveChar` in a different code section (different from core execution logic) more arbitrary. `ReceiveChar` is at the same level as every other action but it's raised from `execute` to a higher level `process_key_bindings`. My simple `bool` meant whether to allow or disallow suppressing keys (not completely true). To keep my comment short, just asking that using an enum maybe wouldn't be a nicer solution?

Reply:
> I feel testing for Action::ReceiveChar in a different code section (different from core execution logic) more arbitrary.

Execution logic is solely about executing actions. If we would print the char manually, that's where it would go. However we shouldn't instruct it to set something that should be done later. There's nothing executed when we're setting a flag.

> My simple bool meant whether to allow or disallow suppressing keys

Actions are not bound to keys. They're arbitrary actions that should be executed. It makes no sense to return if we want to suppress chars or not when executing mouse bindings. So this logic should be handled where it makes sense, at the place where we handle key bindings specifically.

> just asking that using an enum as execute() result maybe wouldn't be a nicer solution?

If execute would ever have any kind of return type, it would likely just indicate success or not. Anything else doesn't make sense.

Reply:
`ReceiveChar` once executes some code in `execute()`, than _independently_ from this it suppresses keys as another distinct side effect. These two are relating to each other therefore should be one unit. What came into my mind was, what if some day `ClearLogNotice` only wants to suppress keys only when really there is some log notice to clear, otherwise it let keys through. (Just now see that it's two distinct key actions.)

Reply:
> ReceiveChar once executes some code in execute(), than independently from this it suppresses keys as another distinct side effect

No it doesn't. `ReceiveChar` doesn't execute any code in the `execute` method

> What came into my mind was, what if some day ClearLogNotice only wants to suppress keys only when really there is some log notice to clear, otherwise it let keys through

First of all, that would be awful for the user when actions suddenly change behavior like that. Second of all, we already leave chars through with `ClearLogNotice`. Thirdly, if changes are required in the future, they should be made in the future, not now. And finally, even if `ClearLogNotice` would want to mingle with suppressing characters we could still do that in the key press handling where it isn't called from other functions.

Reply:
I'm pretty torn on this. There **is** a difference between mouse and keyboard actions in the sense that a keyboard action typically suppresses the underlying default action. This isn't relevant to the mouse. However, I feel like this boils down to an argument for functional style, pass value as bool "flag" in the return type, vs setting some state.

I'm generally a fan of functional styles, but I agree with @chrisduerr that `execute` seems like the wrong place. An `Action` shouldn't care about the surrounding processing in general, and I'd opt to handle `RecieveChar` in a special case within `process_key_bindings`. I haven't played with this in code, so hopefully it makes sense.